server/s3_delete_object_action.cc

```diff
@@ -293,6 +294,39 @@ void S3DeleteObjectAction::delete_metadata_failed() {
   send_response_to_s3_client();
 }
 
+void S3DeleteObjectAction::save_bucket_counters() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  int64_t inc_object_count = 0;
+  int64_t inc_obj_size = 0;
+
+  inc_object_count = -1;
+  inc_obj_size = -(object_metadata->get_content_length());
+
+  S3BucketCapacityCache::update_bucket_capacity(
+      request, bucket_metadata, inc_object_count, inc_obj_size,
+      std::bind(&S3DeleteObjectAction::save_bucket_counters_success, this),
+      std::bind(&S3DeleteObjectAction::save_bucket_counters_failed, this));
+
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Exit\n", __func__);
+}
+
+void S3DeleteObjectAction::save_bucket_counters_success() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Exit\n", __func__);
+  s3_del_obj_action_state = S3DeleteObjectActionState::metadataDeleted;
+  next();
+}
+
+// TODO : how to handle failures to save bucket counters at this stage.
+// Currently just logging error and moving ahead.
```

Comment:
1st way is put this call before metadata deletion.
2nd way -- try roll back here (scary).
3rd way -- let BG service handle it.

I feel 1st way is probably best.

Reply:
To be done in Non-happy path JIRA